packages/x-data-grid-premium/src/hooks/features/rowGrouping/gridRowGroupingUtils.ts

```diff
@@ -79,15 +96,72 @@ const shouldApplyFilterItemOnGroup = (columnField: string, node: GridGroupNode)
 export const filterRowTreeFromGroupingColumns = (
   params: FilterRowTreeFromTreeDataParams,
 ): Omit<GridFilterState, 'filterModel'> => {
-  const { apiRef, rowTree, isRowMatchingFilters, filterModel } = params;
+  const {
+    apiRef,
+    rowTree,
+    isRowMatchingFilters,
+    filterModel,
+    rowGroupingColumnMode,
+    groupingColDef,
+  } = params;
   const filteredRowsLookup: Record<GridRowId, boolean> = {};
   const filteredChildrenCountLookup: Record<GridRowId, number> = {};
   const filteredDescendantCountLookup: Record<GridRowId, number> = {};
   const filterCache = {};
 
+  const getGroupingCriteria = () => {
```

Comment:
It looks like we can call this function only once and use the result, instead of calling it in `shouldFilterGroup` function for every node. WDYT?

Reply:
Also, what's the intetion of this function? I see that it returns null if the grouping column has `leafField`, can you explain what's the purpose of that?

Reply:
`shouldFilterGroup` method predicts where filtering can be applied based on the values in`filterModel` and grouping criteria from `getGroupingCriteria` method
we don't want to filter the group ( i.e, it returns `false`) when any of the following conditions are true :

1- there is a valid `leafField`
2- there are quickFilterValues in filterModel
3-   when `rowGroupingColumnMode`  is `multiple` and when filtering can be applied on multiple grouping columns,  
      it returns `false` when  `groupingField` property in group node from `rowTree` doesn't match the last field in 
     `groupFilterItems`.
             
I hope I'm not missing any edge case to consider.

Reply:
> It looks like we can call this function only once and use the result, instead of calling it in `shouldFilterGroup` function for every node. WDYT?

yes, we can improve here